lint/long-chain_checker.go

```diff
@@ -0,0 +1,121 @@
+package lint
+
+import (
+	"fmt"
+	"go/ast"
+	"strings"
+
+	"github.com/Quasilyte/astcmp"
+)
+
+type longChainChecker struct {
+	ctx *Context
+}
+
+func newLongChainChecker(ctx *Context) Checker {
+	return &longChainChecker{
+		ctx: ctx,
+	}
+}
+
+func (c *longChainChecker) Check(f *ast.File) {
+	for _, decl := range f.Decls {
+		if decl, ok := decl.(*ast.FuncDecl); ok {
+			if decl.Body == nil {
+				continue
+			}
+			for _, stmt := range decl.Body.List {
+				if stmt, ok := stmt.(*ast.SwitchStmt); ok {
+					c.checkSwitch(stmt)
+				}
+			}
+		}
+	}
+}
+
+func (c *longChainChecker) checkSwitch(stmt *ast.SwitchStmt) {
+	exprs := []ast.Expr{}
+	for _, st := range stmt.Body.List {
+		s := st.(*ast.CaseClause)
+		if s.List == nil {
+			continue
+		}
+		exprs = append(exprs, s.List...)
+	}
+	if len(exprs) < 2 {
+		return
+	}
+
+	cp := c.exprToList(exprs[0])
+	for _, e := range exprs[1:] {
+		cp = c.commonPrefix(cp, c.exprToList(e))
+	}
+
+	const n = 3
+
+	if len(cp) > n {
+		c.warn(cp, stmt)
+	}
+}
+
+func (c *longChainChecker) exprToList(expr ast.Expr) []ast.Expr {
```

Comment:
This method is rather hard to understand.
Can you either add comments that describe what (and, more importantly, why) you are doing here
or simplify it? The hardest part is proving it's correct.
It seems like it collects selector expression `X`s, but what if there are non-selector expressions along the way?
Like this: `x.y[i].z()`.

Reply:
the idea is to find the most long prefix, which would fit all expressions.

Reply:
Comment should go to the source file, not here. :)
We are (hopefully) not the only maintainers of this linter, so it should be as accessible for others as possible.